### PR TITLE
FM2-347: Add support for _has for ServiceRequest and Observation

### PR DESCRIPTION
**Pull request by md-at-slashwhy**
## Description of what I changed
This PR adds the [reverse chaining operator (_has)](https://hl7.org/fhir/search.html#has) for TestOrders given the restrictions laid out in the ticket.

## Issue I worked on
see https://issues.openmrs.org/browse/FM2-347

## Checklist: I completed these to help reviewers :)
- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

- [x] I have **added tests** to cover my changes. (If you refactored existing code that was well tested you do not have to add tests)

- [x] I ran `mvn clean package` right before creating this pull request and added all formatting changes to my commit.

- [x] All new and existing **tests passed**.

- [x] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/fhir2/api/search/param/ServiceRequestSearchParams.java
Comment (mherman22) [inline api/src/main/java/org/openmrs/module/fhir2/api/search/param/ServiceRequestSearchParams.java:1]: it's better to use this format we have used for other search params. i.e https://github.com/openmrs/openmrs-module-fhir2/blob/master/api/src/main/java/org/openmrs/module/fhir2/api/search/param/RelatedPersonSearchParams.java

We use this class [here ](https://github.com/openmrs/openmrs-module-fhir2/blob/master/api/src/main/java/org/openmrs/module/fhir2/api/search/param/BaseResourceSearchParams.java)to contain shared search (and search result) parameters that apply to all resources.

**Review by mherman22 — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirServiceRequestDaoImpl.java [hunk 1]
```diff
@@ -89,4 +111,275 @@ private Optional<Criterion> handleDateRange(DateRangeParam dateRangeParam) {
...
+				
+				if (parameterValue != null) {
+					encounterProviderQuery.createAlias("provider", "provider");
+					encounterProviderQuery.add(Restrictions.eq("provider.uuid", parameterValue));
+				}
+				
+				observationQuery.add(propertyIn("encounter", encounterProviderQuery));
+				break;
+			
+			// TODO: add explanation on why these values are not implemented
```
Comment (mherman22) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirServiceRequestDaoImpl.java:264]: i would add this explanation in the method javadoc

File: api/src/main/java/org/openmrs/module/fhir2/api/search/param/ServiceRequestSearchParams.java
Comment (md-at-slashwhy) [reply api/src/main/java/org/openmrs/module/fhir2/api/search/param/ServiceRequestSearchParams.java:1]: Thank you for the hint. I tried to make my changes analogous to [another has PR](https://github.com/openmrs/openmrs-module-fhir2/pull/398/files#diff-2527c9f67b4f5faea9f739e37a2da8739d611cc4ca033903e26b957b6516126a) and didn't notice that a base class has been introduced in the meantime.
I updated the SearchParams and updated the setters and getters for the old `uuid` property to use the id property of the base class.

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirServiceRequestDaoImpl.java (hunk 1 above)
Comment (md-at-slashwhy) [reply api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirServiceRequestDaoImpl.java]: I was unable to determine _why_ these properties couldn't be represented, but added the information to the javadoc.

**Comment by md-at-slashwhy** [comment]
I added additional tests to check failure cases. Would you kindly run the workflow again?

**Comment by itstanany** [comment]
Thanks for this pull request!  It addresses a valuable functionality gap in the OpenMRS FHIR2 module by implementing the _has parameter for reverse chaining related to ServiceRequest and Observation resources.

The description clearly outlines the problem and the proposed solution using the HAPI HasParam class.  The inclusion of unit tests demonstrates a commitment to code quality and ensures the functionality works as expected.

Here are some additional points to consider:

Logging: While the core functionality seems sound, consider if there's a benefit to adding logging statements for _has parameter usage.  These logs could help with debugging and monitoring potential issues related to this specific search parameter.

Documentation:  Double-check if the pull request includes comprehensive documentation for the implemented _has functionality. This documentation should be clear for developers who might use this functionality in the future.  It could include code comments or an update to the FHIR2 module documentation.

Future Considerations: The pull request currently focuses on _has for ServiceRequest and Observation.  Could there be potential future enhancements to support _has for other resource types within the OpenMRS FHIR2 module?  Mentioning this as a future consideration could spark further discussion.

Overall, this pull request effectively addresses the identified issue and represents a positive contribution to the OpenMRS FHIR2 module.  The inclusion of unit tests and clear descriptions strengthen the quality of this contribution.

**Comment by md-at-slashwhy** [comment]
Thank you for the feedback Tanany, glad to hear this feature is valued and that the effort towards quality is noticeable!
I tried adding warning level logs to each known edge and failure case that might produce unexpected results. Adding trace logging for building the query seemed a little too much, as the resulting query can be logged already.
Regarding documentation it's hard for me to judge what might or might not be useful documentation. I tried adding documentation to each method that does not adhere to running standards in the project (as I understood them). But I'd be glad to add more if you see anything that is unclear or might confuse or mislead somebody. Feel free to add comments and I'll try my best to clarify (and refactor) the code to be more understandable.

Regarding future considerations: As this is my first PR in the OpenMRS project we decided to cut the scope of the PR to have an achievable goal. In addition to extending the _has Parameter to more Searches, the current implementation of the _has Parameter does not allow for all the options described in the FHIR spec. For example, the implementation of a _has...:not was something that was explored early during this Tickets development but then removed from the scope. I totally agree that this functionality could be expended much further!

**Comment by md-at-slashwhy** [comment]
Development on the feature (#530) this PR is blocked by seems to have stalled and with no clear information on when this PR might be able to move forward, I can no longer dedicate time to maintaining this PR.  
Feel free to fork the feature and continue development once #530 is done.

**Pull request closed without merging**